pkg/controller/rollout/rollout_bluegreen.go

```diff
@@ -0,0 +1,487 @@
+			klog.Infof("rollout(%s/%s) step(%d) state from(%s) -> to(%s)", c.Rollout.Namespace, c.Rollout.Name,
+				bluegreenStatus.CurrentStepIndex, currentStepStateBackup, bluegreenStatus.CurrentStepState)
+		} else {
+			bluegreenStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			bluegreenStatus.CurrentStepState = v1beta1.CanaryStepStateInit
+			klog.Infof("rollout(%s/%s) step(%d) state from(%s) -> to(%s)", c.Rollout.Namespace, c.Rollout.Name,
+				bluegreenStatus.CurrentStepIndex, currentStepStateBackup, v1beta1.CanaryStepStateInit)
+		}
+		klog.Infof("rollout(%s/%s) canary step from(%d) -> to(%d)", c.Rollout.Namespace, c.Rollout.Name, currentIndexBackup, bluegreenStatus.CurrentStepIndex)
+		return true
+	}
+	return false
+}
+
+// cleanup after rollout is completed or finished
+func (m *blueGreenReleaseManager) doCanaryFinalising(c *RolloutContext) (bool, error) {
+	blueGreenStatus := c.NewStatus.BlueGreenStatus
+	if blueGreenStatus == nil {
+		return true, nil
+	}
+	// rollout progressing complete, remove rollout progressing annotation in workload
+	err := removeRolloutProgressingAnnotation(m.Client, c)
+	if err != nil {
+		return false, err
+	}
+
+	tr := newTrafficRoutingContext(c)
+	// execute steps based on the predefined order for each reason
+	nextStep := nextBlueGreenTask(c.FinalizeReason, blueGreenStatus.FinalisingStep)
+	// if current step is empty, set it with the first step
+	// if current step is end, we just return
+	if len(blueGreenStatus.FinalisingStep) == 0 {
+		blueGreenStatus.FinalisingStep = nextStep
+		blueGreenStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+	} else if blueGreenStatus.FinalisingStep == v1beta1.FinalisingStepTypeEnd {
+		klog.Infof("rollout(%s/%s) finalising process is already completed", c.Rollout.Namespace, c.Rollout.Name)
+		return true, nil
+	}
+	klog.Infof("rollout(%s/%s) Finalising Step is %s", c.Rollout.Namespace, c.Rollout.Name, blueGreenStatus.FinalisingStep)
+
+	var retry bool
+	// the order of steps is maitained by calculating thenextStep
+	switch blueGreenStatus.FinalisingStep {
+	// set workload.pause=false; set workload.partition=0
+	case v1beta1.FinalisingStepTypeBatchRelease:
+		retry, err = finalizingBatchRelease(m.Client, c)
+	// delete batchRelease
+	case v1beta1.FinalisingStepTypeDeleteBR:
+		retry, err = removeBatchRelease(m.Client, c)
+	// restore the gateway resources (ingress/gatewayAPI/Istio), that means
+	// only stable Service will accept the traffic
+	case v1beta1.FinalisingStepTypeGateway:
+		retry, err = m.trafficRoutingManager.RestoreGateway(tr)
+	// restore the stable service
+	case v1beta1.FinalisingStepTypeStableService:
+		retry, err = m.trafficRoutingManager.RestoreStableService(tr)
+	// remove canary service
+	case v1beta1.FinalisingStepTypeRemoveCanaryService:
+		retry, err = m.trafficRoutingManager.RemoveCanaryService(tr)
+	// route all traffic to new version
+	case v1beta1.FinalisingStepTypeRouteAllTraffic:
```

Comment:
consider rename these finalizing steps so that the name can be more differentiating. now it is very hard to tell the difference between FinalisingStepTypeRouteAllTraffic and FinalisingStepTypeGateway, and it is hard to guess the meaning of step FinalisingStepTypeBatchRelease and FinalisingStepTypeGateway, consider rename them to 
 FinalisingStepTypeRestoreWorkload and FinalisingStepTypeRestoreGateway

Reply:
ok!